libnmstate/state.py

```diff
@@ -167,6 +226,49 @@ def merge_interfaces(self, other_state):
             dict_update(other_state.interfaces[name], self.interfaces[name])
             self._ifaces_state[name] = other_state.interfaces[name]
 
+    def merge_route_config(self, current):
+        """
+        Merge routes from current to self.
+        If any changed route referring non-exist interface, create the
+        interface state with name only, in order to trigger the recreating of
+        the interface profile.
+        """
+        # Merge other_routes
+        iface_route_sets = defaultdict(set)
+        current_iface_route_sets = defaultdict(set)
+
+        for route in self.routes:
+            if Route.NEXT_HOP_INTERFACE in route:
+                iface_route_sets[route[Route.NEXT_HOP_INTERFACE]].add(
+                    _Route(route))
+            else:
+                logging.warning("Ignoring the route entry with no "
+                                "next hop interface defined: %s",
+                                route)
+
+        for route in current.routes:
+            iface_route_sets[route[Route.NEXT_HOP_INTERFACE]].add(
```

Comment:
Why are you adding "other"/current to the "self" container?

Reply:
Merging current route into self.

Reply:
Merging other/current routes into desired routes.

Reply:
But a route in the desired may have less properties than the one in current, no?
Also, it is very confusing to read... why not keep each in their own set and then perform operation on all of them... it may be even more efficient.